ndc-client/src/apis/default_api.rs

```diff
@@ -247,37 +218,67 @@ pub async fn schema_get(
 ) -> Result<crate::models::SchemaResponse, Error> {
     let tracer = global::tracer("engine");
     tracer
-        .in_span("schema_get", |ctx| async {
-            let client = &configuration.client;
-
-            let uri = append_path(&configuration.base_path, &["schema"])
-                .map_err(|_| Error::InvalidBaseURL)?;
-            let mut req_builder = client.request(reqwest::Method::GET, uri);
-
-            req_builder = inject_trace_context(req_builder);
+        .in_span("schema_get", |ctx| {
+            async {
+                let client = &configuration.client;
 
-            if let Some(ref user_agent) = configuration.user_agent {
-                req_builder = req_builder.header(reqwest::header::USER_AGENT, user_agent.clone());
-            }
+                let uri = append_path(&configuration.base_path, &["schema"])
+                    .map_err(|_| Error::InvalidBaseURL)?;
+                let mut req_builder = client.request(reqwest::Method::GET, uri);
 
-            // Note: The headers will be merged in to any already set.
-            req_builder = req_builder.headers(configuration.headers.clone());
+                req_builder = inject_trace_context(req_builder);
 
-            let req = req_builder.build()?;
-            let resp = client.execute(req).with_traced_errors().await?;
+                if let Some(ref user_agent) = configuration.user_agent {
+                    req_builder =
+                        req_builder.header(reqwest::header::USER_AGENT, user_agent.clone());
+                }
 
-            let response_status = resp.status();
-            let response_content = resp.json().with_traced_errors().with_context(ctx).await?;
+                // Note: The headers will be merged in to any already set.
+                req_builder = req_builder.headers(configuration.headers.clone());
 
-            if !response_status.is_client_error() && !response_status.is_server_error() {
-                serde_json::from_value(response_content).map_err(Error::from)
-            } else {
-                Err(construct_error(response_status, response_content))
+                let req = req_builder.build()?;
+                execute_request(configuration, req).await
             }
+            .with_context(ctx)
         })
         .await
 }
 
+async fn execute_request<T: DeserializeOwned>(
+    configuration: &configuration::Configuration,
+    request: reqwest::Request,
+) -> Result<T, Error> {
+    async {
+        let client = &configuration.client;
+        let mut resp = client.execute(request).await?;
+        let response_status = resp.status();
+        let response_content = match &configuration.response_size_limit {
+            Some(size_limit) => {
+                let mut size = 0;
+                let mut buf = bytes::BytesMut::new();
```

Comment:
Can we use the `Content-Length` header, if present, to pre-allocate this with the correct capacity?

Reply:
Is there no existing crate we can pull this from? Or even better, put it behind a trait? It doesn't seem to me like this crate needs to be every kind of client for every use case. Maybe even if engine needs a different sort of client, it could just provide its own.

Reply:
I'm not sure what this adds beyond just failing fast if the `Content-Length` exceeds some limit. Isn't this what `.json()` does.

_Edit_: yes, compare with `hyper::body::to_bytes`.

Reply:
> Or even better, put it behind a trait?

A good idea! The `ndc-client` has to be more generic and need not do all the work. I'd prefer to have the custom logic within `v3-engine`. A trait (let's call `NdcClient`) can enable to do this.

> I'm not sure what this adds beyond just failing fast if the Content-Length exceeds some limit

:100: True

Reply:
I found a [`Limited`](https://docs.rs/http-body-util/latest/http_body_util/struct.Limited.html) http body impl in `http_body_util` crate. We could use this to stream the response with a limit on the size. I can't just use it since the `reqwest` restricts accessing inner `res: hyper::Response` of `reqwest::Response` (or) the `reqwest::Response` does not implement `hyper::Body`. 

I feel, the `reqwest` don't want us to bother much about internal (low-level) client details. Now we have either `chunk` or `bytes_stream` methods to incrementally consume the response.